Update Layer guidance (#3040)
grommet/hpe-design-system: aries-site/src/pages/components/layer/index.mdx

### When to use double confirmation

When a user attempts to close out of a layer, the layer closing behavior should always avoid losing user's work.

When data has been edited or changed and the user tries to close the layer, a double confirmation should be shown.

| Layer state                             | Confirmation type   | Notes                                                                                                                                                                                                                      |
| --------------------------------------- | ------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| User has not input or changed any data. | No confirmation     | The user will not lose anything if the layer closes immediately.                                                                                                                                                           |

> | User has input or changed data.         | Double confirmation | The double confirmation should show when the user clicks the "X" or Cancel button, clicks outside of the layer, or presses the Escape key. This short confirmation ensures that user's work won't unintentionally be lost. |

Author: Thank you, to be updated.

### Double confirmation anatomy

> A double confirmation appears as a center layer. This placement puts the confirmation at the center of the user's attention, reinforcing the importance of the dialog's prompt and minimizing distractions from other page content.

Author: Sounds good, I'll says "center modal" and link to the below modal vs non-modal

<Example plain>
  <DoubleConfirmationAnatomy />
</Example>

| Label  | Region       | Purpose                                                              | Required? | Notes                                                                                                      |
| ------ | ------------ | -------------------------------------------------------------------- | --------- | ---------------------------------------------------------------------------------------------------------- |
| **1a** | **Title**    | Explains that data will be discarded when the user closes the layer. | required  | --                                                                                                         |

> | **1b** | **Subtitle** | To clarify that changes will not be saved.                           | required  | --                                                                                                         |

Author: Can you propose some guidance here?

> | **1b** | **Subtitle** | To clarify that changes will not be saved.                           | required  | --                                                                                                         |

Reviewer: On second thought, perhaps a do/don't section would be better here. Thinking something like this: 

![image](https://user-images.githubusercontent.com/86075038/205376215-5425df58-b403-4577-bc58-e5b761ad69b8.png)

> | **1b** | **Subtitle** | To clarify that changes will not be saved.                           | required  | --                                                                                                         |

Author: Great idea

| **2**  | **Footer **  | Contains two actions: "Cancel" and "Discard"                         | required  | The actions should be right-aligned with primary action on the right, to align with center layer guidance. |

> ### Double confirmation example

Reviewer: Disregard if this example is still in a WIP

The "add application" layer here doesn't seem to behave correctly for me. I can't scroll within the layer but I can still scroll through the content behind it. Also looks like the filters button doesn't have content.

Author: Yeah that's a WIP still.

Author: For now, just having the layer open with respect to the whole window. There seems to be a bug with the layer scroll when the layer is restricted to the example container. Planning to debug that separately from this PR.

First, click "Add application" then click the close button in the layer header. Notice how the layer immediately closes.

Next, click "Add application" again and this time type something into the "Title" input. Now, click the close button. Notice how a double confirmation appears.

<Example
  docs="https://v2.grommet.io/layer?theme=hpe#props"
  code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/components/layer/ConfigurationForm.js"
  height={{ min: 'small' }}
  template
  showResponsiveControls={[]}
>
  <ConfigurationForm />
</Example>

> ## Modal vs. non-modal

Author: Layer is the component that drives the layout, elevation, positioning on the screen.

[Modal](https://developer.mozilla.org/en-US/docs/Web/CSS/:modal) is a specific HTML behavior that excludes interaction with outside elements until it is dismissed.

Non-modal means that the outside content can still be interacted with while the layer is open.

Layer is the component. Modal vs. non-modal effects the behavior to dismiss the layer and interactivity of outside content.

Let me know if that clarifies.

> ## Modal vs. non-modal

Author: i added the above clarification (slightly reworded) to the modal vs non-modal section.

> ## Modal vs. non-modal

Reviewer: Yep, that makes sense. Still concerned though that these terms might cause some confusion among folks but again that might just be me.

> ## Modal vs. non-modal

Reviewer: Didn't see your second comment before posting my reply, adding that clarification to that section sounds good to me 👍

<BestPracticeGroup>
  <Example
    height={{ min: 'small' }}
    bestPractice={{
      type: 'do',
      message: `Match the primary action verb to that of the layer's title.`,
    }}
    width="100%"
  >
    <ActionLabelTitle />
  </Example>
  <Example
    height={{ min: 'small' }}
    bestPractice={{
      type: 'dont',
      message: `Don’t use a different verb for the primary action than the layer title's verb.`,
    }}
    width="100%"
  >
    <ActionLabelTitle bestPractice={false} />
  </Example>
</BestPracticeGroup>

## Accessibility

### Layer titles

- Layer headings should be semantically correct.
- Treat a Layer as if it were a stand alone page, with the exception that the first heading within a Layer should be an `h2` or `<Heading level={2} />`.
- All subsequent headings should follow typical [semantic heading structure](/foundation/typography#semantic-usage-of-heading-levels).

### Close button

When using a Layer with `modal={false}`, it is critical that the focus is moved to the "Close" button inside the Layer. Give the close button an `a11yTitle` that provides the user with context that they are in a Layer.

> For example, implementation of this "Close" button for a Layer would be:

Author: When a layer is modal, the focus is already on the layer container so it behaves a little differently and we don't have to explicitly add the focus to the close button. I can revisit the experience for modal and add any necessary guidance/clarification.

      value: 'small',
    },
    {
      label: 7,
      property: 'elevation',
      value: 'large',
    },
  ]}
/>

## Types of layers

Different use cases and content warrant different layer types. Learn more about each type below.

<LayerTypes />

## Layer closing behavior

Depending on if a layer is informational or actionable, the elements and interactions that will close a layer differ.

### Informational

Informational layers do not require any actions from the user, and therefore do not contain any actions.

The following should cause an informational layer to close:

1. Clicking on the close button in the layer's header.
1. Clicking outside of the layer, `onClickOutside`.
1. Pressing the Escape key, `onEsc`.

Learn more about the specifics of [informational center layers](/components/layer/center-layer#informational), [informational side drawer layers](/components/layer/side-drawer-layer#informational), and [informational fullscreen layers](/components/layer/fullscreen-layer#informational).

### Actionable

Actionable layers require the user to either accept/apply the prompt in the layer or cancel out.

The following should cause an actionable layer to close:

> 1. Clicking the close button in the layer's header.

Reviewer 2: To be comprehensive, we should also allow an actionable layer to be closed upon an affirmative action such as 'save' or 'apply.'

Author: Done.

   - The exception to this rule is that actionable center layers should not include a close button in the layer's header. Because center layers generally present confirmations, the user should have to intentionally choose between accepting or cancelling the layer's prompt.
1. Clicking the cancel button in the layer's footer.

> 1. Pressing the Escape key, `onEsc`.

Author: Correct, it also seems to act as a "catch all" for users if they're unsure how to interact with a UI as a universal "get me out" while still being more intentionally pressed than clicking outside of the layer.

> 1. Pressing the Escape key, `onEsc`.

Author: Let me know if some kind of added rationale might be helpful there.

> 1. Pressing the Escape key, `onEsc`.

Author: and if mentioning the "intentionality" on the interaction is suffiicient.

> 1. Pressing the Escape key, `onEsc`.

Reviewer 2: Thanks. I think we can leave as is for now.

Clicking outside of an actionable layer should not close the layer. The user should be intentionally choosing between accepting or cancelling the layer's prompt.

> Learn more about the specifics of [actionable center layers](/components/layer/center-layer#actionable), [actionable side drawer layers](/components/layer/side-drawer-layer#actionable), and [actionable fullscreen layers](/components/layer/fullscreen-layer#actionable).

Reviewer 2: ```suggestion
Use of a double confirmation may be desirable as part of the layer's closing behavior in some instances. See more detail about [when to use double confirmation](#when-to-use-double-confirmation).

Learn more about the specifics of [actionable center layers](/components/layer/center-layer#actionable), [actionable side drawer layers](/components/layer/side-drawer-layer#actionable), and [actionable fullscreen layers](/components/layer/fullscreen-layer#actionable).
```

> ### When to use double confirmation

Reviewer 2: ```suggestion
### Double confirmations

#### When to use double confirmation
```

Thinking that it'd be nice to group all the double confirmation content. The Layer closing behavior is starting to get lengthy and flat otherwise.

> When a user attempts to close out of a layer, the layer closing behavior should always avoid losing user's work.

Reviewer 2: ```suggestion
When closing out of an actionable layer containing editable or configurable information, the behavior should err on preventing a user's unintentional loss of work by presenting a double confirmation.
```

> When data has been edited or changed and the user tries to close the layer, a double confirmation should be shown.

Reviewer 2: ```suggestion
A double confirmation should be shown when data has been edited or changed and the user closes the layer by using the close button, cancel button, or Escape key; a double confirmation is not needed upon affirmative actions like 'save' or 'apply'.
```

| Layer state                             | Confirmation type   | Notes                                                                                                                                                                                         |
| --------------------------------------- | ------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| User has not input or changed any data. | No confirmation     | The user will not lose anything if the layer closes immediately.                                                                                                                              |
| User has input or changed data.         | Double confirmation | The double confirmation should show when the user clicks the "X" or Cancel button, or presses the Escape key. This short confirmation ensures that user's work won't unintentionally be lost. |

### Double confirmation anatomy

> A double confirmation appears as a center [modal](#modal-vs-non-modal). This placement puts the confirmation at the center of the user's attention, reinforcing the importance of the dialog's prompt and minimizing distractions from other page content.

Reviewer 2: 👍

      message:
        'Lead with the single action that needs to be confirmed. In most cases, this is "Discard" followed by a noun of what will be discarded.',
    }}
    width="100%"
  >
    <DoubleConfirmationBestPractice />
  </Example>
  <Example
    height={{ min: 'small' }}
    bestPractice={{
      type: 'dont',
      message: `Don’t include unecessary introductions that create ambiguity around what will happen.`,
    }}
    width="100%"
  >
    <DoubleConfirmationBestPractice bestPractice={false} />
  </Example>
</BestPracticeGroup>

<BestPracticeGroup>
  <Example
    height={{ min: 'small' }}
    bestPractice={{
      type: 'do',
      message: 'Align the primary action verb with the title.',
    }}
    width="100%"
  >
    <DoubleConfirmationBestPractice />
  </Example>
  <Example
    height={{ min: 'small' }}
    bestPractice={{
      type: 'dont',
      message: `Don’t change the primary action to something generic like "Yes". The user should know the consequence of the button by the label alone.`,
    }}
    width="100%"
  >
    <DoubleConfirmationBestPractice bestPractice="badLabel" />
  </Example>
</BestPracticeGroup>

### Double confirmation example

> First, click "Add application" then click the close button in the layer header. Notice how the layer immediately closes.

Reviewer 2: ```suggestion
The example below demonstrates how a double confirmation is only used if the layer's content has been changed or edited. A double confirmation is not used if no changes have been made. This behavior errs on preventing unintentional loss of work that a user has done.

First, click "Add application" then click the close button in the layer header. Notice how the layer immediately closes.
```
